Daemonize hwctl (#596)
canonical/hardware-api: client/README.md

The **Hardware API Client** is the tool to check the certification status of
hardware configurations.

It consists of both the `hwctl-daemon` service and `hwctl` CLI tool.

- `hwctl-daemon`: Rust service for communicating with
  [the API server](../server/) and cache results.

> - `hwctl`: CLI tool to retrieve the certification status and other data.

Reviewer: Again, `hwlib` is still a thing, so I'd say let's leave it in the readme, along with the CLI and daemon

Author: Mmm... but the point is that, now, `hwlib` is not really needed anymore... You just connect to the Varlink socket and get the data. And since it's a protocol with an IDL and reflection support, static languages use code generators (like `varlink-generator` for Rust, or `vali_compiler` for C), and dynamic languages (like python) just connect to the socket and get all the required data from there to build the interface in run time. So... which uses remain for `hwlib`?

Reviewer: It is still published in crates.io, but more generally, `hwlib` is the name of the client library that is the backend for `hwctl` (and now `hwctl-daemon`).

Author: Good point. Fixed. What about now?

- `hwctl`: CLI tool leveraging the `hwlib` library.
- `hwctl-daemon`: Rust service for communicating with
  [the API server](../server/) and cache results.
- `hwctl`: CLI tool to retrieve the certification status and other data.

## Basic Usage

To check the machine certification status, simply run `hwctl` with

> root[^sudo] privileges:

Reviewer: If the `[^sudo]` footnote is not necessary anymore, let's delete it at the bottom of the readme

Author: Good point! Done.

sudo apt-get install hwctl
```

## Service protocol

> The hwctl service is socket-activated, and it uses [Varlink](https://varlink.org/)

Reviewer: Could you put the `[varlink]` link in the bottom of the markdown along with the rest of the links?

```suggestion
The hwctl service is socket-activated, and it uses [Varlink]
```

Author: Ok.

sudo apt-get install hwctl
```

## Service protocol

The hwctl service is socket-activated, and it uses [Varlink] as the underlying protocol.
The protocol is defined in the `bin/com.ubuntu.hwctl.varlink` file. It has two methods:

* GetCertificationStatus(source: CertificationSource, server_url: ?string) -> (state: State)
  It receives a `source` parameter, which can be any of `server`, `cache` or `auto`,
  and an optional (thus, it can be `null`) string with the certification server URL.
  It returns a State object with these fields:

>   * status: `NotSeen`, `Certificated`, `CertifiedImageExists`, `RelatedCertifiedSystemExists` or `Unknown`.

Reviewer: I think this is a typo for `Certified`, right?

```suggestion
  * status: `NotSeen`, `Certified`, `CertifiedImageExists`, `RelatedCertifiedSystemExists` or `Unknown`.
```

Author: Agh! Yes!

  * certified_url: either `null` if there is no applicable URL, or the URL describing either the current
    hardware, or the URL describing the related certified system.
  * available_releases: an array with the certified OS images available for this device, in the case of
    `CertifiedImageExists`. If not, it will be `null`.

>   * valid_cache: a boolean that specifies whether the cache is valid or is old/not-existen/other problem.

Reviewer: ```suggestion
  * valid_cache: a boolean that specifies whether the cache is valid or is old/non-existent/other problem.
```

>   * hardware_mismatch: a boolean that specifies if the current hardware has changed respect to the one

Reviewer: ```suggestion
  * hardware_mismatch: a boolean that specifies if the current hardware has changed in respect to the one
```

    used to get the certified status. Only has meaning if the returned data was obtained from the cache.
  * stale: a boolean that is TRUE if the last connection to the server to update the certified status
    failed (and, thus, the current data is the previous cached one).
  * stale_reason: either `null`, or a string specifying why the last connection to the server failed.
  * source: either `Cache` or `Server`, specifying whether the data shown was obtained fresh from
    the server, or is the cached one.
  * remote_access_enabled: a boolean specifying if the service should refresh the cached data when
    using the `auto` source, or not.
  * server_url: a string with the server URL from which all the shown data was obtained in origin.

* SetRemoteAccess(enabled: bool) -> ()
  It receives a single boolean that changes the `remote_access_enabled` setting inside the service.

For snapped services, the socket path is `/var/snap/hwctl/common/hwctl.varlink`, while for
unconfined services, it is `/run/hwctl/hwctl.varlink`. Clients must always check the former

> first, and only check for the later if the former doesn't exists.

Reviewer: ```suggestion
first, and only check for the latter if the former doesn't exists.
```
